### PR TITLE
Rewrote The Logic in the Ultrasonic Sensor

### PR DESCRIPTION
**Pull request by ddrobner**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 1]
```diff
@@ -26,9 +28,8 @@
         this.ultrasonicSensor = sensor;
         this.time = time;
         this.distance = distance;
-        this.noBalls = false;
-        this.startTime = 0;
 
+        this.firstRun = false; // Not sure if this should be static
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: It should not be static. The `firstRun` state is unique to each poller instance, not shared across all of them.

Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Should it not be initialized to `true`, and then set `false` by the first run when it happens? ie

```
boolean firstRun = true;

task() {
  if (firstRun) {
    this.firstRun = false;
    // do first-run-only things
  }
}
```

**Review by andrewazores — CHANGES_REQUESTED** [review]
firstRun doesn't work as written

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 1 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Yep, that's a mistake

**Comment by andrewazores** [comment]
Still broken. `firstRun`'s value is only set once. You need it to be `true` once and then set `false` after that.

Also that comment about "static" should be removed.

**Comment by ddrobner** [comment]
Oops forgot that. I suppose I should also make it a lifecycle listener so I can reset the value of firstRun when the subsystem is started and stopped.

**Comment by andrewazores** [comment]
That sounds like a good idea too.

**Comment by ddrobner** [comment]
Is it possible to do that without making it an interface?

**Comment by andrewazores** [comment]
Yes, `class Foo implements SomeInterface`

**Comment by ddrobner** [comment]
Actually wait. What I want to do is reset the value when the task is enabled/disabled, not when the lifecycle status is changed. AFAIk there's no way to do that, right?

**Comment by andrewazores** [comment]
```
@Override
public void cancel() {
  super.cancel();
  this.firstRun = true;
}
```

**Comment by ddrobner** [comment]
Ah cool, easier than I thought it would be. Totally didn't think about using `super` in this situation

**Comment by ddrobner** [comment]
Probably would be worth adding that functionality to yggdrasil at some point. Another solution to this problem could be doing a similar thing, except having a method that is run only when the subsystem is started

**Comment by ddrobner** [comment]
Looks like I'm going to have to do that. The method is final so I can't override it

**Comment by andrewazores** [comment]
I somewhat agree and somewhat disagree. Experience over the years so far as proved that needing a "first run" hook like this is rare - this is the first case we've done it out of all of the subsystems we've implemented, dating back to the roots of Yggdrasil in the 2014 robot code. So, adding this hook and others like it when they're rarely used seems like unnecessary complication. Obviously from the example I just gave, for the situations where it is needed, it isn't that hard to add in. And it wouldn't be difficult to do something similar tied to lifecycles either, using standard polymorphism and OOP rather than making Subsystem implementations into do-everything Swiss army knives.

**Comment by andrewazores** [comment]
I can agree with removing `final` from `cancel` and methods like it though.

**Comment by andrewazores** [comment]
FWIW `start` is non-final, so you can do
```
@Override
public void start() {
  this.firstRun = true;
  super.start();
}
```

**Comment by andrewazores** [comment]
This is not as preferable since if the subsystem is still running (wasn't cancelled first) and `run` is called again, then there is a timing window - particularly for very rapidly repeating subsystems - where the existing task can see the `firstRun` flag become reset to true, and then the restarted task will not see it. However this is probably okay for this particular application, it's just a potential source for bugs in a more generic case.

**Comment by ddrobner** [comment]
Yeah I can see that. Seems like a we'll cross that bridge when we come to it kind of thing, like you said.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 2]
```diff
@@ -26,9 +28,8 @@
         this.ultrasonicSensor = sensor;
         this.time = time;
         this.distance = distance;
-        this.noBalls = true;
-        this.startTime = 0;
 
+        this.firstRun = true; // Not sure if this should be static
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Remove this `static` comment. It should not be.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 3]
```diff
@@ -47,24 +48,27 @@ public void setTime(long time) {
...
-        if(this.startTime == 0){
-            this.startTime = System.nanoTime(); // I know about a do while, but this isn't contained within a while loop 
-        }
-        
-        if((this.currentRange > this.distance) && (TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - this.startTime)  > this.time)) {
-            this.noBalls = true;
-        } else if(this.currentRange <= this.distance) {
+    @Override
+    public void task(){
+        if(firstRun) {
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:64]: You still haven't set `firstRun` to `false` within this branch, so `firstRun` will be `true` on every single run of the task.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:64]: Ah missed these two. I did that but then messed up something in git and lost all my changes. Forgot to put those two back in

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:64]: Oh wait it was probably the merge conflict. But whatever

**Review by andrewazores — APPROVED** [review]

**Comment by ddrobner** [comment]
Re-reading what I wrote. It looks like it makes no sense. It's impossible for the start time to be greater than the nanoTime

**Comment by ddrobner** [comment]
I'll fix that

**Comment by ddrobner** [comment]
Pretty sure that last commit I made should have fixed it

**Comment by andrewazores** [comment]
Which "last commit" do you mean?

**Comment by ddrobner** [comment]
On master

**Pull request merged**